File: lowest-common-ancestor-of-a-binary-search-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int is_ancestor(struct TreeNode *root, struct TreeNode *child)
{
    if(root == NULL)
        return 0;
    if(root == child)
        return 1;
    return is_ancestor(root->left,child) || is_ancestor(root->right,child);
}

struct TreeNode* lowestCommonAncestor(struct TreeNode* root, struct TreeNode* p, struct TreeNode* q) {
    if(is_ancestor(root->left,p) && is_ancestor(root->left,q))
        return lowestCommonAncestor(root->left,p,q);
    if(is_ancestor(root->right,p) && is_ancestor(root->right,q))
        return lowestCommonAncestor(root->right,p,q);

    return root;
}

File: python/tree/lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        if not root or root == p or root == q:
            return root

        left = self.lowestCommonAncestor(root.left,p,q)
        right = self.lowestCommonAncestor(root.right,p,q)

        if left and right:
            return root

        return left if left else right
